brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java

```diff
@@ -44,90 +42,54 @@
-      executionAttributes.putAttribute(DEFERRED_ROOT_CONTEXT, maybeDeferredRootSpan.context());
-    } else {
-      executionAttributes.putAttribute(APPLICATION_SPAN, maybeDeferredRootSpan.start());
-    }
-  }
-
-  /**
-   * Before an individual http request happens, may be run multiple times if retries occur
+   * Before an individual http request is finalized. This is only called once per operation, meaning
+   * we can only have one span per operation.
    */
   @Override public SdkHttpRequest modifyHttpRequest(
       Context.ModifyHttpRequest context,
       ExecutionAttributes executionAttributes
   ) {
-    TraceContext maybeDeferredRootSpan = executionAttributes.getAttribute(DEFERRED_ROOT_CONTEXT);
-    Span applicationSpan;
     HttpClientRequest request = new HttpClientRequest(context.httpRequest());
-    if (maybeDeferredRootSpan != null) {
-      Boolean sampled = httpTracing.clientRequestSampler().trySample(request);
-      if (sampled == null) {
-        sampled = httpTracing.tracing().sampler().isSampled(maybeDeferredRootSpan.traceId());
-      }
-      applicationSpan = tracer.toSpan(maybeDeferredRootSpan.toBuilder().sampled(sampled).build());
-      executionAttributes.putAttribute(APPLICATION_SPAN, applicationSpan.start());
-    } else {
-      applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
-    }
+
+    Span span = handler.handleSend(request);
+    executionAttributes.putAttribute(SPAN, span);
 
     String serviceName = executionAttributes.getAttribute(SdkExecutionAttribute.SERVICE_NAME);
     String operation = getAwsOperationNameFromRequestClass(context.request());
-    applicationSpan.name("aws-sdk")
+    span.name("aws-sdk")
         .tag("aws.service_name", serviceName)
         .tag("aws.operation", operation);
 
-    Span span = tracer.newChild(applicationSpan.context());
-    handler.handleSend(request, span);
     span.name(operation).remoteServiceName(serviceName);
-    executionAttributes.putAttribute(CLIENT_SPAN, span);
 
     return request.build();
   }
 
   /**
-   * After individual http response, may run multiple times if retries occur
+   * Before sending an http request. Will be called multiple times in the case of retries.
    */
-  @Override
-  public void beforeUnmarshalling(
-      Context.BeforeUnmarshalling context,
-      ExecutionAttributes executionAttributes
-  ) {
-    Span clientSpan = executionAttributes.getAttribute(CLIENT_SPAN);
-    if (!context.httpResponse().isSuccessful()) {
-      clientSpan.tag("error", context.httpResponse().statusText()
-          .orElse("Unknown AWS service error"));
-    }
-    handler.handleReceive(new HttpClientResponse(context.httpResponse()), null, clientSpan);
-    executionAttributes.putAttribute(CLIENT_SPAN, null);
+  @Override public void beforeTransmission(Context.BeforeTransmission context,
+      ExecutionAttributes executionAttributes) {
+    Span span = executionAttributes.getAttribute(SPAN);
+    span.annotate("ws");
```

Comment:
check null as things can be wonky (someone can delete all attributes)

Reply:
Ah nice point and scary